drivers/metal/metal_device_properties.mm

```diff
@@ -0,0 +1,338 @@
+/* you may not use this file except in compliance with the License.       */
+/* You may obtain a copy of the License at                                */
+/*                                                                        */
+/*     http://www.apache.org/licenses/LICENSE-2.0                         */
+/*                                                                        */
+/* Unless required by applicable law or agreed to in writing, software    */
+/* distributed under the License is distributed on an "AS IS" BASIS,      */
+/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or        */
+/* implied. See the License for the specific language governing           */
+/* permissions and limitations under the License.                         */
+/**************************************************************************/
+
+#include "metal_device_properties.h"
+
+#import <Metal/Metal.h>
+#import <spirv_cross.hpp>
+#import <spirv_msl.hpp>
+
+// Common scaling multipliers
+#define KIBI (1024)
+#define MEBI (KIBI * KIBI)
+
+MTLGPUFamily &operator--(MTLGPUFamily &family) {
+	family = static_cast<MTLGPUFamily>(static_cast<int>(family) - 1);
+	if (family < MTLGPUFamilyApple1) {
+		family = MTLGPUFamilyMetal3;
+	}
+
+	return family;
+}
+
+void MetalDeviceProperties::init_gpu_properties(id<MTLDevice> device) {
+	device_type = RenderingDevice::DEVICE_TYPE_INTEGRATED_GPU;
+	device_vendor = "Apple";
+	device_name = device.name.UTF8String;
+}
+
+void MetalDeviceProperties::init_features(id<MTLDevice> device) {
+	features = { 0 };
+
+	features.highestFamily = MTLGPUFamilyApple1;
+	for (MTLGPUFamily family = MTLGPUFamilyApple9; family >= MTLGPUFamilyApple1; --family) {
+		if ([device supportsFamily:family]) {
+			features.highestFamily = family;
+			break;
+		}
+	}
+
+	features.hostMemoryPageSize = sysconf(_SC_PAGESIZE);
+
+	for (SampleCount sc = SampleCount1; sc <= SampleCount64; sc <<= 1) {
+		if ([device supportsTextureSampleCount:sc]) {
+			features.supportedSampleCounts |= sc;
+		}
+	}
+
+	features.layeredRendering = [device supportsFamily:MTLGPUFamilyApple5];
+	features.multisampleLayeredRendering = [device supportsFamily:MTLGPUFamilyApple7];
+	features.tesselationShader = [device supportsFamily:MTLGPUFamilyApple3];
+	features.imageCubeArray = [device supportsFamily:MTLGPUFamilyApple3];
+	features.quadPermute = [device supportsFamily:MTLGPUFamilyApple4];
+	features.simdPermute = [device supportsFamily:MTLGPUFamilyApple6];
+	features.simdReduction = [device supportsFamily:MTLGPUFamilyApple7];
+
+	features.mslVersionEnum = MTLLanguageVersion1_1;
+
+	if (@available(macOS 11, iOS 14, *)) {
+		features.mslVersionEnum = MTLLanguageVersion2_3;
+	}
+	if (@available(macOS 12, iOS 15, *)) {
+		features.mslVersionEnum = MTLLanguageVersion2_4;
+	}
+	if (@available(macOS 13, iOS 16, *)) {
+		features.mslVersionEnum = MTLLanguageVersion3_0;
+	}
+	if (@available(macOS 14, iOS 17, *)) {
+		features.mslVersionEnum = MTLLanguageVersion3_1;
+	}
```

Comment:
This should be in `MAC_OS_X_VERSION_MAX_ALLOWED` check to allow building on older SDKs. Assuming the implementation require 13.5, but if this is the case, why are there `available` checks for older versions?

Reply:
Fixed 👍🏻